File: folder_size.rb
#!/usr/bin/env ruby
# folder_size.rb
# encoding: UTF-8

require 'find'
require 'optparse'
require 'pathname'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

def human_readable(size)
  units = ['B', 'KB', 'MB', 'GB', 'TB']
  s = size.to_f
  i = 0
  while s >= 1024 && i < units.length - 1
    s /= 1024
    i += 1
  end
  sprintf("%.1f %s", s, units[i])
end

def get_folder_size(dir, recursive, depth, max_depth, exclude_hidden, verbose, size_map)
  total = 0
  begin
    entries = Dir.entries(dir) - ['.', '..']
  rescue Errno::EACCES
    puts colorize("Permission denied: #{dir}", :red) if verbose
    size_map[dir] = 0
    return 0
  end

  entries.each do |name|
    next if exclude_hidden && name.start_with?('.')
    full = File.join(dir, name)
    if File.file?(full)
      sz = File.size(full)
      total += sz
      puts "  #{full}: #{human_readable(sz)}" if verbose
    elsif File.directory?(full) && recursive && (max_depth == 0 || depth < max_depth)
      sub_map = {}
      sub_total = get_folder_size(full, recursive, depth+1, max_depth, exclude_hidden, verbose, sub_map)
      total += sub_total
      sub_map.each { |k, v| size_map[k] = v }
    end
  end
  size_map[dir] = total
  total
end

options = {
  path: '.',
  recursive: true,
  human: false,
  sort: false,
  top: 0,
  depth: 0,
  exclude_hidden: false,
  verbose: false
}

OptionParser.new do |opts|
  opts.banner = "Usage: folder_size.rb [options] [path]"
  opts.on("-p", "--path DIR", "Path") { |v| options[:path] = v }
  opts.on("-r", "--recursive", "Recursive") { options[:recursive] = true }
  opts.on("-h", "--human-readable", "Human readable") { options[:human] = true }
  opts.on("-s", "--sort", "Sort by size") { options[:sort] = true }
  opts.on("-t", "--top N", Integer, "Show top N") { |v| options[:top] = v }
  opts.on("-d", "--depth N", Integer, "Max depth") { |v| options[:depth] = v }
  opts.on("--exclude-hidden", "Exclude hidden") { options[:exclude_hidden] = true }
  opts.on("-v", "--verbose", "Verbose") { options[:verbose] = true }
  opts.on("-h", "--help", "Help") { puts opts; exit }
end.parse!

root = File.expand_path(options[:path] || ARGV[0] || '.')
unless Dir.exist?(root)
  puts colorize("Error: directory not found", :red)
  exit 1
end

size_map = {}
get_folder_size(root, options[:recursive], 0, options[:depth], options[:exclude_hidden],
                options[:verbose], size_map)

items = size_map.to_a
items.sort! { |a, b| b[1] <=> a[1] } if options[:sort]
if options[:top] > 0 && options[:top] < items.size
  items = items[0...options[:top]]
end

max_path_len = items.map { |p, _| p.length }.max || 0
items.each do |path, size|
  size_str = options[:human] ? human_readable(size) : "#{size} B"
  color = if size > 1024*1024*1024
            :red
          elsif size > 1024*1024
            :yellow
          else
            :green
          end
  puts "#{colorize(size_str.rjust(12), color)}  #{path}"
end

File: folder_size.js
// folder_size.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const stat = promisify(fs.stat);
const readdir = promisify(fs.readdir);

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

function humanReadable(size) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let s = size;
    let i = 0;
    while (s >= 1024 && i < units.length - 1) {
        s /= 1024;
        i++;
    }
    return `${s.toFixed(1)} ${units[i]}`;
}

async function getFolderSize(dir, recursive, depth, maxDepth, excludeHidden, verbose) {
    const result = {};
    let total = 0;

    try {
        const entries = await readdir(dir, { withFileTypes: true });
        for (const entry of entries) {
            if (excludeHidden && entry.name.startsWith('.')) continue;
            const full = path.join(dir, entry.name);
            if (entry.isFile()) {
                const st = await stat(full);
                total += st.size;
                if (verbose) console.log(`  ${full}: ${humanReadable(st.size)}`);
            } else if (entry.isDirectory() && recursive && (maxDepth === 0 || depth < maxDepth)) {
                const sub = await getFolderSize(full, recursive, depth+1, maxDepth, excludeHidden, verbose);
                for (const [k, v] of Object.entries(sub)) {
                    result[k] = v;
                }
                // Суммируем размер подпапки
                let subTotal = 0;
                for (const v of Object.values(sub)) subTotal += v;
                total += subTotal;
            }
        }
    } catch (err) {
        if (verbose) console.log(colorize(`Permission denied: ${dir}`, 'red'));
        return { [dir]: 0 };
    }
    result[dir] = total;
    return result;
}

async function main() {
    const args = require('minimist')(process.argv.slice(2), {
        string: ['p'],
        boolean: ['r', 'h', 's', 'exclude-hidden', 'v'],
        alias: { p: 'path', r: 'recursive', h: 'human-readable', s: 'sort', t: 'top', d: 'depth' },
        default: { p: '.', r: true, t: 0, d: 0 }
    });
    const root = path.resolve(args.p);
    const recursive = args.r;
    const human = args.h;
    const sortFlag = args.s;
    const top = args.t || 0;
    const maxDepth = args.d || 0;
    const excludeHidden = args['exclude-hidden'] || false;
    const verbose = args.v || false;

    try {
        const st = await stat(root);
        if (!st.isDirectory()) throw new Error('Not a directory');
    } catch (err) {
        console.log(colorize(`Error: ${err.message}`, 'red'));
        process.exit(1);
    }

    const sizeMap = await getFolderSize(root, recursive, 0, maxDepth, excludeHidden, verbose);
    let items = Object.entries(sizeMap);
    if (sortFlag) {
        items.sort((a, b) => b[1] - a[1]);
    }
    if (top > 0 && top < items.length) {
        items = items.slice(0, top);
    }

    for (const [p, size] of items) {
        let sizeStr = human ? humanReadable(size) : `${size} B`;
        let color = 'green';
        if (size > 1024*1024*1024) color = 'red';
        else if (size > 1024*1024) color = 'yellow';
        console.log(`${colorize(sizeStr.padStart(12), color)}  ${p}`);
    }
}

main().catch(err => {
    console.log(colorize(`Error: ${err.message}`, 'red'));
    process.exit(1);
});
